app/lib/common/widgets/md_editor_with_preview.dart

```diff
@@ -0,0 +1,111 @@
+  }) : super(key: key);
+
+  @override
+  ConsumerState<ConsumerStatefulWidget> createState() =>
+      _MdEditorWithPreviewState();
+}
+
+class _MdEditorWithPreviewState extends ConsumerState<MdEditorWithPreview> {
+  bool _showPreview = false;
+  final TextEditingController _textCtr = TextEditingController();
+
+  @override
+  Widget build(BuildContext context) {
+    final controller = widget.controller ?? _textCtr;
+    return Expanded(
+      child: Padding(
+        padding: const EdgeInsets.symmetric(vertical: 10),
+        child: Stack(
+          children: [
+            _showPreview
+                ? FormField(
+                    builder: (x) => Expanded(
+                      child: Container(
+                        decoration: BoxDecoration(
+                          border: Border.all(
+                            width: 3,
+                          ),
+                          borderRadius: BorderRadius.circular(5),
+                        ),
+                        child: Padding(
+                          padding: const EdgeInsets.all(10),
+                          child: Consumer(
+                            builder: (context, ref, child) => ref
+                                .watch(markdownProvider(controller.text))
+                                .when(
+                                  data: (text) => RenderHtml(text: text),
+                                  error: (error, stackTrace) =>
+                                      Text('Parsing markdown failed: $error'),
+                                  loading: () => const Text('Parsing ...'),
+                                ),
+                          ),
+                        ),
+                      ),
+                    ),
+                    validator: widget
+                        .validator, // make sure we still have the validator in the tree
+                  )
+                : TextFormField(
+                    decoration: InputDecoration(
+                      hintText: widget.hintText,
+                      labelText: widget.labelText,
+                      border: OutlineInputBorder(
+                        borderRadius: BorderRadius.circular(5),
+                      ),
+                    ),
+                    textAlignVertical: TextAlignVertical.top,
+                    controller: controller,
+                    expands: true,
+                    minLines: null,
+                    maxLines: null,
+                    keyboardType: TextInputType.multiline,
+                    validator: widget.validator,
+                    onChanged: (String? value) {
+                      widget.onChanged ?? (value);
+                    },
+                  ),
+            Positioned(
+              right: 10,
+              bottom: 10,
+              child: Tooltip(
+                message: 'Toggle preview',
+                child: IconButton(
+                  onPressed: () => setState(() => _showPreview = !_showPreview),
```

Comment:
Using ```setState``` is all good for widgets with not much deep tree. Makes code simpler. But alternatively we would keep states in provider for easier access and testing.

Reply:
Yeah, for simple toggles I do not quite agree to be honest ... a provider for a simple true/false flag that _literally_ only switches a particular view from one to the others seems like a lot of overhead...